### PR TITLE
Make local and binary file locations relative

### PR DESCRIPTION
**Pull request by cdrage**
Makes the local and binary file locations relative instead of absolute
paths within the .odo/config.yaml file.

Closes: https://github.com/openshift/odo/issues/1552

### CODE REVIEW

**Comment by cdrage** [comment]
WIP:

 - [ ] Need to fix `odo watch`
 - [ ] Add binary support
 - [ ] Tests

This is my current implementation @kadel it worked well until I started troubleshooting `odo watch`. To be honest, we have quite a lot code that needs to be cleaned up / refactored within the CLI... Where I thought I fixed the path, it would appear somewhere else again!

File: pkg/odo/cli/component/create.go [hunk 1]
```diff
@@ -107,60 +111,65 @@ func NewCreateOptions() *CreateOptions {
...
+		componentSourceType = config.GIT
+	}
+	co.componentSettings.SourceType = &componentSourceType
+
+	// Here we set the correct source path for each type
+	switch componentSourceType {
+
+	// --binary
+	case config.BINARY:
 		cPath, err := util.GetAbsPath(co.componentBinary)
```
Comment (kadel) [inline pkg/odo/cli/component/create.go]: `GetAbsPath` looks suspicious, bath to the binary should be relative

File: pkg/odo/cli/component/push.go [hunk 2]
```diff
@@ -76,12 +75,16 @@ func (po *PushOptions) Complete(name string, cmd *cobra.Command, args []string)
 	po.localConfig = conf
 	po.sourceType = conf.LocalConfig.GetSourceType()
 
+	glog.V(0).Infof("SourceLocation: %s", po.localConfig.GetSourceLocation())
+
 	// Get SourceLocation here...
-	po.sourcePath, err = conf.GetOSSourcePath()
+	po.sourcePath, err = conf.GetOSSourcePath(po.componentContext)
```
Comment (kadel) [inline pkg/odo/cli/component/push.go]: `po.componentContext` can be a relative path, and by default, it is (`./`) so before using it for `GetOsSourcePath` will have to convert it to the full abolute path. `util.GetAbsPath` might be handy for that.

**Comment by kadel** [comment]
> This is my current implementation @kadel it worked well until I started troubleshooting `odo watch`. To be honest, we have quite a lot code that needs to be cleaned up / refactored within the CLI... Where I thought I fixed the path, it would appear again somewhere else..

what is broken in `odo watch` I've tried to do some quick tests, and it seems to work for me

**Comment by mohammedzee1000** [comment]
Seems to be going good so far

**Comment by girishramnani** [comment]
Any regression tests for this?

File: pkg/config/config.go [hunk 3]
```diff
@@ -574,7 +574,10 @@ func GetSrcType(ctStr string) (SrcType, error) {
 // GetOSSourcePath corrects the current sourcePath depending on local or binary configuration,
 // if Git has been passed, we simply return the source location from LocalConfig
 // this will get the correct source path whether on Windows, macOS or Linux.
-func (lci *LocalConfigInfo) GetOSSourcePath() (path string, err error) {
+//
+// This function also takes in the current working directory + context directory in order
+// to correctly retrieve WHERE the source is located..
+func (lci *LocalConfigInfo) GetOSSourcePath(componentContext string) (path string, err error) {
```
Comment (anmolbabu) [inline pkg/config/config.go]: why is componentContext required?
Can we use filepath.Dir(filepath.Dir(lci.Filename)) instead ?
Ref: https://github.com/openshift/odo/blob/1c3081ce6406bb99ba5027cade553895efd4da6a/pkg/config/config.go#L92

**Comment by kadel** [comment]
/retest 
```
 ✗  the server has asked for the client to provide credentials (get builds.build.openshift.io)
```

**Comment by kadel** [comment]
Haven't looked much into the code, but tested it  with a few examples and it looks good 

/approve

File: pkg/odo/cli/component/create.go [hunk 4]
```diff
@@ -329,12 +342,14 @@ func (co *CreateOptions) Complete(name string, cmd *cobra.Command, args []string
...
+		// Above code is for INTERACTIVE mode
 	} else {
-		err = co.setCmpSourceAttrs()
+		// Else if NOT using interactive / UI
+		err = co.setComponentSourceAttributes()
 		if err != nil {
 			return err
 		}
-		err = co.setCmpName(args)
+		err = co.setComponentName(args)
```
Comment (mohammedzee1000) [inline pkg/odo/cli/component/create.go]: +1 for name expansion

File: pkg/odo/cli/component/create.go [hunk 5]
```diff
@@ -107,60 +113,66 @@ func NewCreateOptions() *CreateOptions {
 	return &CreateOptions{}
 }
 
-func (co *CreateOptions) setCmpSourceAttrs() (err error) {
+func (co *CreateOptions) setComponentSourceAttributes() (err error) {
```
Comment (mohammedzee1000) [inline pkg/odo/cli/component/create.go]: +1

**Comment by mohammedzee1000** [comment]
/hold
@cdrage do you want to update any check lists?

**Comment by kadel** [comment]
/test e2e-core

```
•Running command: odo project list -o json
{"kind":"List","apiVersion":"odo.openshift.io/v1alpha1","metadata":{},"items":[{"kind":"Project","apiVersion":"odo.openshift.io/v1alpha1","metadata":{"name":"json-test","creationTimestamp":null},"spec":{"apps":["myapp"]},"status":{"active":true}},{"kind":"Project","apiVersion":"odo.openshift.io/v1alpha1","metadata":{"name":"myproject","creationTimestamp":null},"spec":{"apps":["myapp"]},"status":{"active":false}},{"kind":"Project","apiVersion":"odo.openshift.io/v1alpha1","metadata":{"name":"odo2-1555342753","creationTimestamp":null},"spec":{"apps":["myapp"]},"status":{"active":false}}]}
Result: 
 stdout: {"kind":"List","apiVersion":"odo.openshift.io/v1alpha1","metadata":{},"items":[{"kind":"Project","apiVersion":"odo.openshift.io/v1alpha1","metadata":{"name":"json-test","creationTimestamp":null},"spec":{"apps":["myapp"]},"status":{"active":true}},{"kind":"Project","apiVersion":"odo.openshift.io/v1alpha1","metadata":{"name":"myproject","creationTimestamp":null},"spec":{"apps":["myapp"]},"status":{"active":false}},{"kind":"Project","apiVersion":"odo.openshift.io/v1alpha1","metadata":{"name":"odo2-1555342753","creationTimestamp":null},"spec":{"apps":["myapp"]},"status":{"active":false}}]}

 stderr: 
 exitcode: 0 

------------------------------
• Failure [0.347 seconds]
odojsonoutput odo machine readable output [It] should be able to list the projects 
/go/src/github.com/openshift/odo/tests/e2e/json_test.go:63

  Expected
      <bool>: false
  to be true

  /go/src/github.com/openshift/odo/tests/e2e/json_test.go:67
```


/test e2e-scenarios

```
level=error msg="1 error occurred:"
level=error msg="\t* module.vpc.aws_vpc.new_vpc: 1 error occurred:"
level=error msg="\t* aws_vpc.new_vpc: Error creating VPC: VpcLimitExceeded: The maximum number of VPCs has been reached."
level=error msg="\tstatus code: 400, request id: b073ccd4-62c1-4000-9e06-4e5896183d10"
```

**Comment by amitkrout** [comment]
Rebase please. Json error fixed in another pr.

**Comment by kadel** [comment]
> * [ ]  Need to fix `odo watch` ignoring files
> * [x]  Add binary support
> * [ ]  Tests

@cdrage are those items addressed in this PR? If not we can open separate issues for those.

**Comment by kadel** [comment]
> > * [ ]   Need to fix `odo watch` ignoring files
> > * [x]   Add binary support
> > * [ ]   Tests
> 
> @cdrage are those items addressed in this PR? If not we can open separate issues for those.

#1630 and #1631 were opened to track those two things 

/hold cancel

**Pull request merged**